pkg/handlers/scripts/install.sh

```diff
@@ -1,116 +1,353 @@
-	echo "{{ if .MoveToPath }}Installing{{ else }}Downloading{{ end }} $USER/$PROG $RELEASE ($URL)..."
-	#enter tempdir
-	mkdir -p $TMP_DIR
-	cd $TMP_DIR
-	if [[ $FTYPE = ".gz" ]]; then
-		which gzip > /dev/null || fail "gzip is not installed"
-		#gzipped binary
-		NAME="${PROG}_${OS}_${ARCH}.gz"
-		GZURL="$GH/releases/download/$RELEASE/$NAME"
-		#gz download!
-		bash -c "$GET $URL" | gzip -d - > $PROG || fail "download failed"
-	elif [[ $FTYPE = ".tar.gz" ]]; then
-		#check if archiver progs installed
-		which tar > /dev/null || fail "tar is not installed"
-		which gzip > /dev/null || fail "gzip is not installed"
-		bash -c "$GET $URL" | tar zxf - || fail "download failed"
-	elif [[ $FTYPE = ".zip" ]]; then
-		which unzip > /dev/null || fail "unzip is not installed"
-		bash -c "$GET $URL" > tmp.zip || fail "download failed"
-		unzip -o -qq tmp.zip || fail "unzip failed"
-		rm tmp.zip || fail "cleanup failed"
-	elif [[ $FTYPE = "" ]]; then
-		bash -c "$GET $URL" > "{{ .Program }}_${OS}_${ARCH}" || fail "download failed"
-	else
-		fail "unknown file type: $FTYPE"
+	[[ ! -z "${OS+x}" ]] || fail "could not determine the OS"
+
+	case $(uname -m) in
+		"amd64" | "x86_64")
+			ARCH="amd64"
+			;;
+		"arm64" | "aarch64")
+			ARCH="arm64"
+			;;
+		"arm")
+			ARCH="arm"
+			;;
+		"i386")
+			ARCH="386"
+			;;
+		*)
+			fail "unsupported architecture $(uname -m)"
+	esac
+	[[ ! -z "${ARCH+x}" ]] || fail "could not determine the architecture"
+
+	# Check for the current OS + arch combination in the available assets.
+	# NOTE: the case statements are built by the templating engine by ranging
+	#   over the set of assets and creating an OS_ARCH case that assigns that
+	#   asset's URL and file type.
+	case "${OS}_${ARCH}" in
+	  {{ range .Assets }}
+		{{ .OS }}_{{ .Arch }})
+			URL="{{ .URL }}"
+			FTYPE="{{ .Type }}"
+			;;
+		{{ end }}
+		*)
+			fail "No asset found for platform ${OS}-${ARCH}"
+			;;
+	esac
+	[[ ! -z "${URL+x}" || ! -z "${FTYPE+x}" ]] || fail "could not find a valid release URL for ${OS} ${ARCH}"
+
+	# Check for needed utilities.
+	command -v curl &> /dev/null || fail "curl not installed"
+	command -v find &> /dev/null || fail "find not installed"
+	command -v xargs &> /dev/null || fail "xargs not installed"
+	command -v sort &> /dev/null || fail "sort not installed"
+	command -v tail &> /dev/null || fail "tail not installed"
+	command -v cut &> /dev/null || fail "cut not installed"
+	command -v du &> /dev/null || fail "du not installed"
+
+	# Check that the assets can be extracted.
+	case "${FTYPE}" in
+		".gz")
+			command -v gzip &>/dev/null || fail "gzip is not installed"
+			;;
+		".tar.gz")
+			command -v tar &>/dev/null || fail "tar is not installed"
+			;;
+		".zip")
+			command -v unzip &>/dev/null || fail "zip is not installed"
+			;;
+		"")
+			;;
+		*)
+			fail "unsupported file type ${FTYPE}"
+	esac
+
+	# Check if the install directory needs to be prompted for and exists.
+	[[ ! -z "${OUT_DIR:+x}" ]] || prompt_install_dir
```

Comment:
I would much rather have this formatted as an if statement
```suggestion
	if [[ -z "${OUT_DIR:+x}" ]]; then
	  prompt_install_dir
	fi
```

Reply:
the same is true of the various `|| fail ...` invocations, but that's at least a lot more self contained whereas here we're intending to continue